substrate/primitives/runtime/src/traits/transaction_extension/mod.rs

```diff
@@ -0,0 +1,629 @@
+	};
+	($call:ty ; $context:ty ; prepare $( $rest:tt )*) => {
+		fn prepare(
+			self,
+			_val: Self::Val,
+			_origin: &$crate::traits::OriginOf<$call>,
+			_call: &$call,
+			_info: &$crate::traits::DispatchInfoOf<$call>,
+			_len: usize,
+			_context: & $context,
+		) -> Result<Self::Pre, $crate::transaction_validity::TransactionValidityError> {
+			Ok(Default::default())
+		}
+		impl_tx_ext_default!{$call ; $context ; $( $rest )*}
+	};
+	($call:ty ; $context:ty ;) => {};
+}
+
+/// Information about a [`TransactionExtension`] for the runtime metadata.
+pub struct TransactionExtensionMetadata {
+	/// The unique identifier of the [`TransactionExtension`].
+	pub identifier: &'static str,
+	/// The type of the [`TransactionExtension`].
+	pub ty: MetaType,
+	/// The type of the [`TransactionExtension`] additional signed data for the payload.
+	// TODO: Rename "implicit"
+	pub additional_signed: MetaType,
+}
+
+#[impl_for_tuples(1, 12)]
+impl TransactionExtensionBase for Tuple {
+	for_tuples!( where #( Tuple: TransactionExtensionBase )* );
+	const IDENTIFIER: &'static str = "Use `metadata()`!";
+	for_tuples!( type Implicit = ( #( Tuple::Implicit ),* ); );
+	fn implicit(&self) -> Result<Self::Implicit, TransactionValidityError> {
+		Ok(for_tuples!( ( #( Tuple.implicit()? ),* ) ))
+	}
+	fn weight() -> Weight {
+		let mut weight = Weight::zero();
+		for_tuples!( #( weight = weight.saturating_add(Tuple::weight()); )* );
+		weight
+	}
+	fn metadata() -> Vec<TransactionExtensionMetadata> {
+		let mut ids = Vec::new();
+		for_tuples!( #( ids.extend(Tuple::metadata()); )* );
+		ids
+	}
+}
+
+#[impl_for_tuples(1, 12)]
+impl<Call: Dispatchable, Context> TransactionExtension<Call, Context> for Tuple {
+	for_tuples!( where #( Tuple: TransactionExtension<Call, Context> )* );
+	for_tuples!( type Val = ( #( Tuple::Val ),* ); );
+	for_tuples!( type Pre = ( #( Tuple::Pre ),* ); );
+
+	fn validate(
+		&self,
+		origin: <Call as Dispatchable>::RuntimeOrigin,
+		call: &Call,
+		info: &DispatchInfoOf<Call>,
+		len: usize,
+		context: &mut Context,
+		self_implicit: Self::Implicit,
+		inherited_implication: &impl Encode,
+	) -> Result<
+		(ValidTransaction, Self::Val, <Call as Dispatchable>::RuntimeOrigin),
+		TransactionValidityError,
+	> {
+		let valid = ValidTransaction::default();
+		let val = ();
+		let following_explicit_implications = for_tuples!( ( #( &self.Tuple ),* ) );
+		let following_implicit_implications = self_implicit;
+
+		for_tuples!(#(
+			// Implication of this pipeline element not relevant for later items, so we pop it.
+			let (_item, following_explicit_implications) = following_explicit_implications.pop_front();
+			let (item_implicit, following_implicit_implications) = following_implicit_implications.pop_front();
+			let (item_valid, item_val, origin) = {
+				let implications = (
+					// The first is the implications born of the fact we return the mutated
+					// origin.
+					inherited_implication,
```

Comment:
I would actually put this one last in the `implications` tuple, it seems to make more sense for nested situations:

for pipeline: `( ( A1 , A2 ) , B , C )`
* currently implications for `A1` are `B_explicit, B_implicit, C_exp, C_imp, A2_exp, A2_imp`.
* if we put last it would be: `A2_exp, A2_imp, B_explicit, B_implicit, C_exp, C_imp`.

Reply:
I think the order matters as long as it's consistent. Although it's technically possible to deserialize the following implication with some associated type in an extension, I don't think there's a valid use case for accessing that information this that can't be solved more elegantly in another way.